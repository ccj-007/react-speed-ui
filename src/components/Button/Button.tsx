import React, { FC, ButtonHTMLAttributes, AnchorHTMLAttributes, useContext } from 'react';
import classNames from 'classnames';
import { ConfigContext } from '../Config-Provider/ConfigProvider';

//先明确需求，定义类型
export type ButtonSize = 'lg' | 'sm';
export type ButtonType = 'primary' | 'success' | 'warning' | 'danger' | 'link';

interface BaseButtonProps {
  /** 样式命名隔离 */
  prefixCls?: string;
  /** 组件类名 */
  className?: string;
  /**设置 Button 的尺寸 */
  size?: ButtonSize;
  /**是否禁用*/
  disabled?: boolean;
  /**设置 Button 的类型 */
  btnType?: ButtonType;
  children?: React.ReactNode;
  href?: string;
  htmlType?: string
}

type NativeButtonProps = BaseButtonProps & ButtonHTMLAttributes<HTMLElement>;
// a 链接属性
type AnchorButtonProps = BaseButtonProps & AnchorHTMLAttributes<HTMLElement>;
export type ButtonProps = Partial<NativeButtonProps & AnchorButtonProps>

/**
 * Button  按钮组件
 */
const Button: FC<ButtonProps> = props => {
  const { className, size, disabled, btnType = 'primary', prefixCls: customizePrefixCls, children, href, htmlType, ...restProps } = props;

  const configInfo = useContext(ConfigContext);
  let prefixCls = configInfo.getPrefixCls("button", customizePrefixCls);

  const classes = classNames(prefixCls, className, {
    [`${prefixCls}-${btnType}`]: btnType,
    [`${prefixCls}-${size}`]: size,
    [`${configInfo.prefixCls}-dark`]: configInfo.isDark, //暗黑主题
    [`${prefixCls}-disabled`]: btnType === 'link' && disabled,
  });

  //自定义组件主题配置
  let themeStyles = {};
  if (configInfo.themeOptions.theme === 'custom' && btnType && btnType !== 'link') {
    themeStyles = {
      backgroundColor: configInfo.themeOptions.button[btnType],
    };
  }

  if (btnType === 'link' && href) {
    return (
      <a data-testid='button' className={classes} href={href} {...restProps}>
        {children}
      </a>
    );
  } else {
    return (
      <button data-testid='button' disabled={disabled} style={themeStyles} className={classes}  {...restProps} >
        {children}
      </button>
    );
  }
};

Button.defaultProps = {
  btnType: 'primary',
};

Button.displayName = 'Button'

export default Button;
