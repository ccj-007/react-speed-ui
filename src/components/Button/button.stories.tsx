import React from "react";
import { action } from "@storybook/addon-actions";
import Button from "./button";

export default {
	title: "通用/Button",
	component: Button,
};

const defaultButtonTpl = () => (
	<Button onClick={action("clicked")}>default button</Button>
);

const buttonWithSizeTpl = (args: any) => (
	<>
		<Button size={args.size}>{args.children}</Button>
		<Button size="sm">small button</Button>
	</>
);

const buttonWithTypeTpl = (args: any) => (
	<>
		<Button btnType={args.btnType}>{args.children}</Button>
		<Button btnType="danger">danger button</Button>
		<Button btnType="link" href="https://google.com">
			link button
		</Button>
	</>
);

export const defaultButton: any = defaultButtonTpl.bind({});
defaultButton.args = {};

export const buttonWithSize: any = buttonWithSizeTpl.bind({});
buttonWithSize.args = {
	size: "lg",
	children: "large button",
};

export const buttonWithType: any = buttonWithTypeTpl.bind({});
buttonWithType.args = {
	btnType: "primary",
	children: "primary button",
};
