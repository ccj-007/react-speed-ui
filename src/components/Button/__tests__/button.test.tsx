import { render } from '@testing-library/react';
import Button from '../Button';
import "@testing-library/jest-dom/extend-expect";

describe('test button component', () => {
  it('should render a <Button/> components', () => {
    const { asFragment } = render(
      <>
        <Button >default button</Button>
        <Button size="lg">small button</Button>
        <Button size="sm">small button</Button>
        <Button btnType="success">success button</Button>
        <Button btnType="primary">primary button</Button>
        <Button btnType="warning">warning button</Button>
        <Button btnType="danger">danger button</Button>
        <Button btnType="link" href="https://google.com">
          link button
        </Button>
      </>
    );
    expect(asFragment()).toMatchSnapshot();
  })

  it('should use speed-ui classNames', () => {
    const { container } = render(
      <>
        <Button size="lg">small button</Button>
      </>
    );
    expect(container.firstChild).toHaveClass('speed-button speed-button-lg')
  })

  it("should be disabled", () => {
    const { getByTestId } = render(<Button disabled />);
    expect(getByTestId("button")).toBeDisabled();
  });


  it("if button type is link so tagName is a", () => {
    const { getByTestId } = render(<Button btnType='link' href='www.baidu.com' />);
    expect(getByTestId("button").tagName).toEqual('A');
  });
});
