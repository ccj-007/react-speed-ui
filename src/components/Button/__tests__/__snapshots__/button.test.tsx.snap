// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test button component should render a <Button/> components 1`] = `
<DocumentFragment>
  <button
    class="speed-button speed-button-primary"
    data-testid="button"
  >
    default button
  </button>
  <button
    class="speed-button speed-button-primary speed-button-lg"
    data-testid="button"
  >
    small button
  </button>
  <button
    class="speed-button speed-button-primary speed-button-sm"
    data-testid="button"
  >
    small button
  </button>
  <button
    class="speed-button speed-button-success"
    data-testid="button"
  >
    success button
  </button>
  <button
    class="speed-button speed-button-primary"
    data-testid="button"
  >
    primary button
  </button>
  <button
    class="speed-button speed-button-warning"
    data-testid="button"
  >
    warning button
  </button>
  <button
    class="speed-button speed-button-danger"
    data-testid="button"
  >
    danger button
  </button>
  <a
    class="speed-button speed-button-link"
    data-testid="button"
    href="https://google.com"
  >
    link button
  </a>
</DocumentFragment>
`;
